ffi/rtc.lua

```diff
@@ -0,0 +1,196 @@
+    end
+
+    if re == 0 then
+        if enabled then
+            self._wakeup_scheduled = true
+        else
+            self._wakeup_scheduled = false
+            self._wakeup_scheduled_ptm = nil
+        end
+        return true
+    else
+        return nil, re, err
+    end
+end
+
+--[[
+Unset wakeup alarm.
+--]]
+function RTC:unsetWakeupAlarm()
+    self:setWakeupAlarm(-1, false)
+    self._wakeup_scheduled = false
+    self._wakeup_scheduled_ptm = nil
+end
+
+--[[--
+Get wakealarm as set by us.
+--]]
+function RTC:getWakeupAlarm()
+    return self._wakeup_scheduled_ptm
+end
+
+--[[--
+Get RTC wakealarm from system.
+--]]
+function RTC:getWakeupAlarmSys()
+    local wake = ffi.new("struct rtc_wkalrm")
+
+    local err, re
+    local rtc0 = C.open("/dev/rtc0", C.O_RDONLY)
+    if rtc0 == -1 then
+        err = ffi.string(C.strerror(ffi.errno()))
+        print("getWakeupAlarm open /dev/rtc0", rtc0, err)
+    end
+    re = C.ioctl(rtc0, C.RTC_WKALM_RD, wake)
+    if re == -1 then
+        err = ffi.string(C.strerror(ffi.errno()))
+        print("getWakeupAlarm ioctl RTC_WKALM_RD", re, err)
+    end
+    re = C.close(rtc0)
+    if re == -1 then
+        err = ffi.string(C.strerror(ffi.errno()))
+        print("getWakeupAlarm close /dev/rtc0", re, err)
+    end
+
+    if wake ~= -1 then
+        local t = ffi.new("time_t[1]")
+        t[0] = C.time(nil)
+        local tm = ffi.new("struct tm") -- luacheck: ignore
+        tm = C.gmtime(t)
+        tm.tm_sec = wake.time.tm_sec
+        tm.tm_min = wake.time.tm_min
+        tm.tm_hour = wake.time.tm_hour
+        tm.tm_mday = wake.time.tm_mday
+        tm.tm_mon = wake.time.tm_mon
+        tm.tm_year = wake.time.tm_year
+        return tm
+    end
+end
+
+--[[--
+Checks if the alarm we set matches the system alarm as well as the current time.
+--]]
+function RTC:validateWakeupAlarmByProximity(task_alarm_epoch, proximity)
+    -- In principle alarm time and current time should match within a second,
+    -- but let's be absurdly generous and assume anything within 30 is a match.
+    proximity = proximity or 30
+
+    local alarm = self:getWakeupAlarm()
+    local alarm_epoch
+    local alarm_sys = self:getWakeupAlarmSys()
+    local alarm_sys_epoch
+
+    -- this seems a bit roundabout
+    local current_time = ffi.new("time_t[1]")
+    current_time[0] = C.time(nil)
+    local current_time_epoch = C.mktime(C.gmtime(current_time))
+
+    if not (alarm and alarm_sys) then return end
+
+    alarm_epoch = C.mktime(alarm)
+    alarm_sys_epoch = C.mktime(alarm_sys)
+
+    print("validateWakeupAlarmByProximity", task_alarm_epoch, alarm_epoch, alarm_sys_epoch, current_time_epoch)
+
+    -- If our stored alarm and the system alarm don't match, we didn't set it.
+    if not (alarm_epoch == alarm_sys_epoch) then return end
+
+    -- If our stored alarm and the provided task alarm don't match,
+    -- we're not talking about the same task. This should never happen.
```

Comment:
Unless 2 features/plugins set each an alarm. This will happen to the oldest plugin that did, no?
(if task_alarm_epoch should be the same used to set the alarm, and that the plugins should store so they can provide it to this function later?)

Reply:
It could happen if a rogue plugin avoided the WakeupMgr:addTask/removeTask mechanism, in which case the check could prevent WakeupMgr from running its top task. (However, I haven't currently implemented that in front.)